File: ios/react-native-bluetooth-cross-platform/NetworkCommunicator.swift
import Foundation
import UIKit
import Underdark

@objc (NetworkCommunicator)
public class NetworkCommunicator: TransportHandler, MessageEncoder, MessageDecoder {
  
  // delimiters
  private var displayDelimeter: String = "$%#";
  private var typeDelimeter: String = "%$#";
  private var deviceDelimeter: String = "$#%";
  private var advertiseTimer: Timer? = nil
  private let deviceId: String = UIDevice.current.identifierForVendor?.uuidString ?? ""
  private let displayName: String = UIDevice.current.name
  private var type: User.PeerType = User.PeerType.OFFLINE
  
  override open func initTransport(_ kind: String, inType: User.PeerType) {
    super.initTransport(kind, inType: inType)
    self.type = inType
    initTimer()
  }
  func initTimer() {
    DispatchQueue.main.async(execute: {
      if self.advertiseTimer == nil {
        self.advertiseTimer = Timer.scheduledTimer(timeInterval: 0.5, target: self, selector: #selector(self.broadcastType), userInfo: nil, repeats: true)
        RunLoop.main.add(self.advertiseTimer!, forMode: RunLoopMode.defaultRunLoopMode)
      }
    })
  }
  
  open override func stopTransport() {
    if advertiseTimer != nil {
      advertiseTimer?.invalidate()
      advertiseTimer = nil
    }
  }
  
  // HANDLE RECIEVING NEW FRAME
  override open func transport(_ transport: UDTransport, link: UDLink, didReceiveFrame frameData: Data) {
    super.transport(transport, link: link, didReceiveFrame: frameData);
    let message = getMessage(frameData: frameData) ?? ""
    let name = getDisplayName(frameData: frameData) ?? ""
    let id = getDeviceId(frameData: frameData) ?? ""
    var user: User? = nil
    
    switch message {
    case "advertiserbrowser":
      user = User(inLink: link, inId: id, inConnected: false, peerType: .ADVERTISER_BROWSER, name: name)
      checkForNewUsers(user!)
      return
    case "advertiser":
      user = User(inLink: link, inId: id, inConnected: false, peerType: .ADVERTISER, name: name)
      checkForNewUsers(user!)
      return
    case "browser":
      user = User(inLink: link, inId: id, inConnected: false, peerType: .BROWSER, name: name)
      checkForNewUsers(user!)
      return
    case "invitation":
      user = findUser(id)
      if user != nil {
        self.sendEvent(withName: "receivedInvitation", body: user!.getJSUser("invitation"))
      }
      return
    case "accepted":
      user = findUser(id)
      if user != nil {
        user!.connected = true
        informConnected(user: user!)
        self.sendEvent(withName: "connectedToUser", body: user!.getJSUser("connected"))
      }
      return
    case "connected":
      user = findUser(id)
      if user != nil {
        user?.connected = true
        self.sendEvent(withName: "connectedToUser", body: user?.getJSUser("connected"))
      }
      return
    case "disconnected":
      user = findUser(id)
      if user != nil {
        user?.connected = false
        self.sendEvent(withName: "lostUser", body: user!.getJSUser("lost peer"))
      }
      break
    default:
      user = findUser(id)
      if user != nil {
        self.sendEvent(withName: "messageReceived", body: user?.getJSUser(message))
      }
      return
    }
  }
  
  // MARK: MESSAGE ENCODER PROTOCOOL
  open func sendMessage(message: String, userId:String) {
    if let user = findUser(userId) {
      let data = "\(displayName)\(displayDelimeter)\(self.type.rawValue)\(typeDelimeter)\(deviceId)\(deviceDelimeter)\(message)".data(using: String.Encoding.utf8)
      user.link.sendFrame(data!)
    }
  }
  open func sendMessage(message: String, link:UDLink) {
    let data = "\(displayName)\(displayDelimeter)\(self.type.rawValue)\(typeDelimeter)\(deviceId)\(deviceDelimeter)\(message)".data(using: String.Encoding.utf8)
    link.sendFrame(data!)
  }
  open func informConnected(user: User) {
    self.sendMessage(message: "connected", link: user.link)
  }
  open func informDisonnected(user: User) {
    sendMessage(message: "disconnected", link: user.link)
    user.connected = false
    self.sendEvent(withName: "lostUser", body: user.getJSUser("lost peer"))
    
  }
  open func informAcceptedInvite(user: User) {
    sendMessage(message: "accepted", link: user.link)
  }
  open func inviteUser(user: User) {
    sendMessage(message: "invitation", link: user.link)
  }
  open func broadcastType() {
    for i in 0..<links.count {
      sendMessage(message: self.type.rawValue, link: links[i])
    }
  }
  // MARK: MESSAGE DECODER PROTOCOOL
  open func getDisplayName(frameData: Data)-> String? {
    let str: String = String(data: frameData, encoding: String.Encoding.utf8) ?? ""
    if let endIndex: Int = str.getIndexOf(displayDelimeter) {
      let displayName = str.substring(with: str.startIndex..<str.characters.index(str.startIndex, offsetBy: endIndex))
      return displayName
    }
    return nil
  }
  open func getType(frameData: Data)-> String? {
    let str: String = String(data: frameData, encoding: String.Encoding.utf8) ?? ""
    if let startIndex: Int = str.getIndexOf(displayDelimeter) {
      if let endIndex: Int = str.getIndexOf(typeDelimeter) {
        return str.substring(with: str.characters.index(str.startIndex, offsetBy: startIndex)..<str.characters.index(str.startIndex, offsetBy: endIndex))
      }
    }
    return nil
  }
  open func getDeviceId(frameData: Data)-> String? {
    let str: String = String(data: frameData, encoding: String.Encoding.utf8) ?? ""
    if let startIndex: Int = str.getIndexOf(typeDelimeter) {
      if let endIndex: Int = str.getIndexOf(deviceDelimeter) {
        let deviceId = str.substring(with: str.characters.index(str.startIndex, offsetBy: startIndex)..<str.characters.index(str.startIndex, offsetBy: endIndex))
        return deviceId
      }
    }
    return nil
  }
  open func getMessage(frameData: Data)-> String? {
    let str: String = String(data: frameData, encoding: String.Encoding.utf8) ?? ""
    if let startIndex: Int = str.getIndexOf(deviceDelimeter) {
      let start = str.index(str.startIndex, offsetBy: startIndex + deviceDelimeter.characters.count)
      let message = str.substring(with: start..<str.endIndex)
      return message
    }
    return nil
  }
  
  // MARK: Utility functions
  open func findUser(_ id: String) -> User? {
    for user in nearbyUsers {
      if user.deviceId == id {
        return user
      }
    }
    return nil
  }
  fileprivate func checkForNewUsers(_ user: User) {
    for i in 0..<nearbyUsers.count {
      if nearbyUsers[i].deviceId == user.deviceId && nearbyUsers[i].mode != user.mode {
        nearbyUsers[i].mode = user.mode;
        return;
      }
      return;
    }
    nearbyUsers.append(user)
    self.sendEvent(withName: "detectedUser", body: user.getJSUser("new user"))
  }
  
  override open func supportedEvents() -> [String]! {
    return ["lostUser","detectedUser", "messageReceived", "connectedToUser", "receivedInvitation"]
  }
}

File: android/src/main/java/com/rctunderdark/NetworkCommunicator.java
package com.rctunderdark;

import android.bluetooth.BluetoothAdapter;
import android.provider.Settings;
import android.provider.Settings.Secure;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.io.UnsupportedEncodingException;
import java.util.Timer;
import java.util.TimerTask;

import io.underdark.transport.Link;
import io.underdark.transport.Transport;

public class NetworkCommunicator extends TransportHandler implements MessageDecoder, MessageEncoder {

    // delimiters
    private String displayDelimeter = "$%#";
    private String typeDelimeter = "%$#";
    private String deviceDelimeter = "$#%";
    private String deviceID = Settings.Secure.getString(context.getContentResolver(), Secure.ANDROID_ID);
    private String displayName;
    private Timer broadcastTimer = null;
    private Boolean isRunning = false;
    private User.PeerType type = User.PeerType.OFFLINE;

    // INITIALIAZATION
    public NetworkCommunicator(ReactApplicationContext reactContext) {
        super(reactContext);
        if (BluetoothAdapter.getDefaultAdapter() != null) {
            displayName = BluetoothAdapter.getDefaultAdapter().getName();
        }
    }

    @Override
    public void transportLinkDidReceiveFrame(Transport transport, Link link, byte[] frameData) {
        super.transportLinkDidReceiveFrame(transport, link, frameData);
        String message = getMessage(frameData);
        String id = getDeviceId(frameData);
        if(id.equals(this.deviceID)) {
            return;
        }
        User.PeerType type = getType(frameData);
        String displayName = getDisplayName(frameData);
        User user = null;
        switch (message) {
            case "advertiserbrowser":
                user = new User(id, displayName, link, false, type);
                checkForNewUser(user);
                return;
            case "advertiser":
                user = new User(id, displayName, link, false, type);
                checkForNewUser(user);
                return;
            case "browser":
                user = new User(id, displayName, link, false, type);
                checkForNewUser(user);
                return;
            case "invitation":
                user = findUser(id);
                if (user != null) {
                    context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                            .emit("receivedInvitation", user.getJSUser());
                }
                return;
            case "accepted":
                user = findUser(id);
                if (user != null) {
                    user.connected = true;
                    informConnected(user);
                }
                return;
            case "connected":
                user = findUser(id);
                if (user != null) {
                    user.connected = true;
                }
                return;
            case "disconnected":
                user = findUser(id);
                if (user != null) {
                    user.connected = false;
                }
                WritableMap map = user.getJSUser();
                map.putString("message", "lost peer");
                context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                        .emit("lostUser", map);
                return;
            default:
                user = findUser(id);
                if (user != null){
                    WritableMap messageMap = user.getJSUser();
                    messageMap.putString("message", message);
                    context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                            .emit("messageReceived", messageMap);
                }
        }
    }

    @Override
    public void initTransport(String kind, User.PeerType inType) {
        super.initTransport(kind, inType);
        this.type = inType;
        broadcastType();
    }

    @Override
    public void stopTransport() {
        if (broadcastTimer != null) {
            broadcastTimer.cancel();
        }
        broadcastTimer = null;
        isRunning = false;
    }

    @Override
    public String getDisplayName(byte[] frame) {
        try {
            String str = new String(frame, "UTF-8");
            int displayEnd = str.indexOf(displayDelimeter);
            return str.substring(0, displayEnd);
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public User.PeerType getType(byte[] frame) {
        try {
            String str = new String(frame, "UTF-8");
            int typeStart = str.indexOf(displayDelimeter);
            int typeEnd = str.indexOf(typeDelimeter);
            String strType = str.substring(typeStart, typeEnd);
            return User.getPeerValue(strType);
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public String getDeviceId(byte[] frame) {
        try {
            String str = new String(frame, "UTF-8");
            int displayStart = str.indexOf(typeDelimeter) + deviceDelimeter.length();
            int displayEnd = str.indexOf(deviceDelimeter);
            return str.substring(displayStart, displayEnd);
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public String getMessage(byte[] frame) {
        try {
            String str = new String(frame, "UTF-8");
            int messageStart = str.indexOf(deviceDelimeter);
            if(messageStart != -1) {
                return str.substring(messageStart + deviceDelimeter.length(), str.toCharArray().length);
            }
            return "index out of range";
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public void informConnected(User user) {
        user.connected = true;
        sendMessage("connected", user.deviceId);
        context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit("connectedToUser", user.getJSUser());
    }

    @Override
    public void informAccepted(User user) {
        user.connected = true;
        sendMessage("accepted", user.deviceId);
        context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit("connectedToUser", user.getJSUser());
    }

    @Override
    public void informDisconnected(User user) {
        user.connected = false;
        sendMessage("disconnected", user.deviceId);
        context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit("lostUser", user.getJSUser());
    }

    @Override
    public void informAcceptedInvite(User user) {
        user.connected = true;
        sendMessage("accepted", user.deviceId);
        context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit("connectedToUser", user.getJSUser());
    }

    @Override
    public void inviteUser(User user) {
        sendMessage("invitation", user.deviceId);
    }

    @Override
    public void broadcastType() {
        if(!isRunning) {
            // creating timer task, timer
            TimerTask tasknew = new TimerTask() {
                @Override
                public void run() {
                    isRunning = true;
                    for(int i = 0; i < links.size(); ++i)
                        sendMessage(User.getStringValue(type), links.elementAt(i));
                }
            };
            broadcastTimer = new Timer();
            broadcastTimer.schedule(tasknew, 0, 1000);
        }
    }

    @Override
    public void sendMessage(String message, Link link) {
        byte[] frame = displayName.concat(displayDelimeter).concat(User.getStringValue(this.type).concat(typeDelimeter).concat(deviceID).concat(deviceDelimeter).concat(message)).getBytes();
        link.sendFrame(frame);
    }

    @Override
    public void sendMessage(String message, String id) {
        User user = findUser(id);
        if(user != null) {
            byte[] frame = displayName.concat(displayDelimeter).concat(User.getStringValue(this.type).concat(typeDelimeter).concat(deviceID).concat(deviceDelimeter).concat(message)).getBytes();
            user.link.sendFrame(frame);
        }
    }

    // UTILITY
    public User findUser(String id) {
        for (int i = 0; i < nearbyUsers.size(); ++i) {
            if (nearbyUsers.elementAt(i).deviceId.contains(id)) {
                return nearbyUsers.elementAt(i);
            }
        }
        return null;
    }
    private void checkForNewUser(User user) {
        if(findUser(user.deviceId) != null) {
            return;
        }
        nearbyUsers.add(user);
        context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit("detectedUser", user.getJSUser());
    }
}
